walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,118 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*; 
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> mandatoryTimeRanges = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> optionalTimeRanges = new ArrayList<TimeRange>();
+    // if meeting duration is longer than a day, it is not possible to hold meeting
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return mandatoryTimeRanges;
+    }
+    // initialize time range with whole day
+    mandatoryTimeRanges.add(TimeRange.WHOLE_DAY);
+    optionalTimeRanges.add(TimeRange.WHOLE_DAY);
+    Collection<String> MandatoryAttendants = request.getAttendees();
+    Collection<String> optionalAttendants = request.getOptionalAttendees();
+    for(Event e: events) {
+      boolean checkMandatoryAttendee = MandatoryAttendants.size() == 0 ? false : true;
+      int numOptionalUnavailable = 0;
+      for(String attendee: e.getAttendees()) {
+        // if one of the attendees of this event is one of the people we're scheduling this meeting for, then find any time conflicts and remove them
+        if (checkMandatoryAttendee && MandatoryAttendants.contains(attendee)) {
+            handleEventConflict(mandatoryTimeRanges, e, request);
+            checkMandatoryAttendee = false;
+        }
+        if (optionalAttendants.contains(attendee)) {
+            numOptionalUnavailable++;
+        }
+      }
+      //if all optional attendees are attending this event, then mark this time slot as unavailable for any optional attendees
+      if (optionalAttendants.size() != 0 && numOptionalUnavailable == optionalAttendants.size()) {
+          handleEventConflict(optionalTimeRanges, e, request);
+      }
+    }
+
+    // See if there are any valid overlaps between the mandatory time ranges and optional time ranges
+    ArrayList<TimeRange> overlapTimeRanges = new ArrayList<TimeRange>();
+    int mandatoryIndex = 0;
+    int optionalIndex = 0;
+    while (mandatoryIndex < mandatoryTimeRanges.size() && optionalIndex < optionalTimeRanges.size()) {
```

Comment:
My algorithm for modifying the timeranges list should ensure that they're being split and inserted in an ordered manner. If you go down a bit you'll see the insertion part when I add to the timeranges. Maybe there is a bug that I don't know of?

Reply:
One good side note here is that when there are implicit/not immediately obvious implications (e.g. time ranges will be sorted), it would be good to add it as a comment to make it explicit to make things clearer.